### ISSUE TITLE
Add `MultiStepBatchNorm` with cupy backend

### ISSUE THREAD

**Issue by fangwei123456** [fangwei123456/spikingjelly#157]
`BatchNorm` is also a stateful module, and has different outputs when using in step-by-step and layer-by-layer mode. So, we need to define a `MultiStepBatchNorm` module.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from spikingjelly.clock_driven.functional import seq_to_ann_forward, multi_step_forward

bn1 = nn.BatchNorm2d(4)
bn2 = nn.BatchNorm2d(4)

x = torch.rand([8, 2, 4, 8, 8])

bn1.train()
bn2.train()

y1 = seq_to_ann_forward(x, bn1)

y2 = multi_step_forward(x, bn2)

with torch.no_grad():
    print((y1 - y2).abs().max())

print(bn1.running_mean, bn1.running_var)
print(bn2.running_mean, bn2.running_var)
```

```
tensor(0.3313)
tensor([0.0495, 0.0502, 0.0514, 0.0493]) tensor([0.9084, 0.9085, 0.9082, 0.9081])
tensor([0.2820, 0.2851, 0.2952, 0.2823]) tensor([0.4778, 0.4792, 0.4771, 0.4772])
```

**Comment by tmasquelier** [comment]
Hi Wei,
I wasn't aware of this discrepancy!
I guess it is due to the running mean and var, isn't it?
In eval mode, the discrepancy disappears, right?
Same thing if track_running_stats=False?

**Comment by fangwei123456** [comment]
> In eval mode, the discrepancy disappears, right?

Yes, this will only affect training because of running mean and var:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from spikingjelly.clock_driven.functional import seq_to_ann_forward, multi_step_forward

bn = nn.BatchNorm2d(4)


x = torch.rand([8, 2, 4, 8, 8])

# recode a random mean and var
bn.train()
multi_step_forward(torch.rand_like(x), bn)


bn.eval()
y1 = seq_to_ann_forward(x, bn)

y2 = multi_step_forward(x, bn)

with torch.no_grad():
    print((y1 - y2).abs().max())
```

The outputs are `tensor(0.)`.


> Same thing if track_running_stats=False?

If so, then `seq_to_ann_forward` and `multi_step_forward` still have different behavior.

**Comment by fangwei123456** [comment]
Consequently,  we can still use layer-by-layer in training to accelerate and step-by-step in inference. I will run an experiment to compare the performance bewteen `seq_to_ann_forward(x, bn)` and `multi_step_forward(x, bn)`. @tmasquelier

**Comment by tmasquelier** [comment]
>>Same thing if track_running_stats=False?
>If so, then seq_to_ann_forward and multi_step_forward still have different behavior.

Any idea why? I guess the set over which the mean and std are computed is different.

>Consequently, we can still use layer-by-layer in training to accelerate and step-by-step in inference

Agreed. In any case, after training, we can always absorb the BN layer into the preceding (stateless) conv layer.

**Comment by fangwei123456** [comment]
>Any idea why? I guess the set over which the mean and std are computed is different.

When `track_running_stats=False`, BN will use mean and var of the inputs.
When we use `seq_to_ann_forward`, `[T, N, C, H, W]` will be regarded as `[TN, C, H, W]`. Then mean and var for `i`-th channel is calculated by `[TN, i, H, W]`.

When we use `multi_step_forward`, Then mean and var for `i`-th channel in `t`-th step is calculated by `[t, N, i, H, W]`.

**Comment by fangwei123456** [comment]
![image](https://user-images.githubusercontent.com/38352109/146370703-17038e67-0d7f-44c8-941d-3b0bd80da139.png)
![image](https://user-images.githubusercontent.com/38352109/146370857-b58e92b0-6b56-4e61-a5bf-751540e2f2ae.png)

I use the network from https://spikingjelly.readthedocs.io/zh_CN/latest/clock_driven_en/14_classify_dvsg.html . `dvsg_logs_msbn` is BN used with `multi_step_forward`. The results show that using BN in `seq_to_ann_forward` has better accuracy. @tmasquelier

**Comment by tmasquelier** [comment]
>When track_running_stats=False, BN will use mean and var of the inputs.
When we use seq_to_ann_forward, [T, N, C, H, W] will be regarded as [TN, C, H, W]. Then mean and var for i-th channel are calculated by [TN, i, H, W].

>When we use multi_step_forward, Then mean and var for i-th channel in t-th step are calculated by [t, N, i, H, W].

Very clear, thanks!
So I guess the estimates with seq_to_ann_forward are more stable because there are T times more datapoints.
It's probably the reason why seq_to_ann_forward was more accurate in your experiment.
